### ISSUE TITLE
cubeb doesn't link on windows in debug if linking with libcpmtd

### ISSUE THREAD

**Issue by mutexlox-signal** [mozilla/cubeb-rs#92]
I'm trying to build cubeb in windows along with a debug version of my app, which also uses windows APIs directly.

For a release build this works fine, but on a debug build of my app, cubeb appears to still use release values, causing a link failure.

Evidently, cubeb_wasapi (and others, e.g. cubeb_mixer, cubeb_log, cubeb_resampler, and cubeb_utils) **statically** links against libcpmt, at least when built via cubeb-rs's build.rs file, so standard approaches of using `/NODEFAULTLIB` and manually specifying the desired libcpmtd do not change what cubeb has pulled in.

Is there a supported way to ask cubeb to use the debug build of libcpmtd?

Error excerpt:

```
note: libcubeb_sys-ca42eb6a204bb3ed.rlib(cubeb_wasapi.obj) : error LNK2038: mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in libcpmtd.lib(excptptr.obj)
          libcubeb_sys-ca42eb6a204bb3ed.rlib(cubeb_wasapi.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MT_StaticRelease' doesn't match value 'MTd_StaticDebug' in libcpmtd.lib(excptptr.obj)
          libcubeb_sys-ca42eb6a204bb3ed.rlib(cubeb_mixer.obj) : error LNK2038: mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in libcpmtd.lib(excptptr.obj)
          libcubeb_sys-ca42eb6a204bb3ed.rlib(cubeb_mixer.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MT_StaticRelease' doesn't match value 'MTd_StaticDebug' in libcpmtd.lib(excptptr.obj)
          libcubeb_sys-ca42eb6a204bb3ed.rlib(cubeb_log.obj) : error LNK2038: mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in libcpmtd.lib(excptptr.obj)
          libcubeb_sys-ca42eb6a204bb3ed.rlib(cubeb_log.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MT_StaticRelease' doesn't match value 'MTd_StaticDebug' in libcpmtd.lib(excptptr.obj)
          libcubeb_sys-ca42eb6a204bb3ed.rlib(cubeb_resampler.obj) : error LNK2038: mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in libcpmtd.lib(excptptr.obj)
          libcubeb_sys-ca42eb6a204bb3ed.rlib(cubeb_resampler.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MT_StaticRelease' doesn't match value 'MTd_StaticDebug' in libcpmtd.lib(excptptr.obj)
          libcubeb_sys-ca42eb6a204bb3ed.rlib(cubeb_utils.obj) : error LNK2038: mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in libcpmtd.lib(excptptr.obj)
          libcubeb_sys-ca42eb6a204bb3ed.rlib(cubeb_utils.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MT_StaticRelease' doesn't match value 'MTd_StaticDebug' in libcpmtd.lib(excptptr.obj)
```

note that I've filed this here because I am not sure whether this is a cubeb or a cubeb-rs issue, but it seems at least plausible that it has to do with the build configuration in cubeb-rs.

**Comment by mutexlox-signal** [comment]
More details: It seems like cubeb is compiling with `/MT` ([multithreaded static version of the library](https://learn.microsoft.com/en-us/cpp/build/reference/md-mt-ld-use-run-time-library?view=msvc-170)) even if build.rs specifies CMAKE_BUILD_TYPE of `Debug`.

This makes it impossible to link with a library that uses an incompatible library (in this case, libcpmtd -- specified by /MTd).

**Comment by padenot** [comment]
@kinetiknz, do you know what we should do here?

**Comment by mutexlox-signal** [comment]
Just wanted to see if there was any chance either of you could take a look at this :)

**Comment by mutexlox-signal** [comment]
I believe that https://github.com/mozilla/cubeb/pull/798 should address the issue.

**Comment by kinetiknz** [comment]
Fixed by https://github.com/mozilla/cubeb/pull/798, we'll need to update the cubeb import to pick this up.

**Comment by kinetiknz** [comment]
Unfortunately this [breaks the debug build](https://github.com/mozilla/cubeb-rs/actions/runs/10894043122/job/30230072538) for `systest`.  `ctest2` uses `cc-rs` to build `systest` -- it looks like `cc-rs` [doesn't support](https://github.com/rust-lang/cc-rs/blob/c0b3eae28300fc8b36eb3a0b519ef4083f1a129e/src/lib.rs#L1926) linking with the debug run-time library.

**Comment by mutexlox-signal** [comment]
Might be able to use the `/NODEFAULTLIB:libcmtd.lib` option when linking to get around that.

**Comment by mutexlox-signal** [comment]
That would be in [build.rs](https://github.com/mozilla/cubeb-rs/blob/master/cubeb-sys/build.rs#L71), i think, something like:

```
println!("cargo:rustc-link-arg=/NODEFAULTLIB:libcmtd.lib");
```

when linking in debug mode.

**Comment by kinetiknz** [comment]
IIUC, in debug builds we want to use libcmtd and avoid the non-debug version.  Since Rust 1.79, the runtime is specified with `/defaultlib` and can be overridden with `/nodefaultlib` (prior to 1.79 it was handled in a different way via the `libc` crate, and couldn't be overridden).

To get `systest` building, emitting `println!("cargo:rustc-link-arg=/nodefaultlib:msvcrt");` in `systest/build.rs` seems to be sufficient.

Unfortunately building the tests is broken in a similar way; adding a `build.rs` with the same` /nodefaultlib` works for cubeb-backend and cubeb-core, but cubeb-api still fails with a different set of linker errors suggesting one of the other runtime libraries (ucrt, vcruntime) is not linked correctly.  Trying to link the debug versions of those explicitly replaces that set of errors with the original one indicating a conflict between libcmtd and another runtime.

**Comment by mutexlox-signal** [comment]
> IIUC, in debug builds we want to use libcmtd and avoid the non-debug version

Yes, typo -- sorry about that.

Where's the other runtime come from when trying to link the debug version of the tests? what's the full error there?

**Comment by mutexlox-signal** [comment]
Nevermind, reproduced this locally. I'll take a bit of a look

**Comment by mutexlox-signal** [comment]
https://github.com/mozilla/cubeb-rs/pull/94 should fix this.

**Comment by mutexlox-signal** [comment]
(at least, it does locally)

**Comment by kinetiknz** [comment]
Fixed by #94
